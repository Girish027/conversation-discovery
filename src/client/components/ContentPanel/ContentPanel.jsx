import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { routeNames } from 'utils/RouteHelper';
import ConnectedRunsView from './RunsView';
import ConnectedConversationsView from './ConversationsView';
import ConnectedTopicDistributionView from './TopicDistributionView';
import PreviewManager from '../../utils/PreviewManager';

export default class ContentPanel extends PureComponent {
  componentDidMount() {
    PreviewManager.init();
  }

  render() {
    const { routeName } = this.props;

    switch (routeName) {
      case routeNames.HOME_PAGE:
      case routeNames.DISCOVER_INTENTS:
        return (<ConnectedRunsView />);
      case routeNames.DISTRIBUTION_GRAPH:
        return (<ConnectedTopicDistributionView />);
      case routeNames.TOPIC_REVIEW:
        return (<ConnectedConversationsView />);
      default:
        return (<ConnectedRunsView />);
    }
  }
}

ContentPanel.propTypes = {
  routeName: PropTypes.string.isRequired,
};
