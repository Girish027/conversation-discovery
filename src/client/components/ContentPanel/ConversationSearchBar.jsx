import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { SearchIcon, Xmark, TextField } from '@tfs/ui-components';
import IconButton from '../IconButton/index';
import { shouldHandle } from '../../utils/KeyboardUtils';
import { searchBarStyle } from '../../styles/index';

class ConversationSearchBar extends Component {
  constructor(props) {
    super(props);

    this.style = {
      paddingTop: '12px',
      paddingLeft: '0px',
      marginTop: '-12px',

      container: {
        marginRight: '0px',
        width: '350px',
      }
    };

    this.state = {
      value: '',
    };

    this.onSearch = this.onSearch.bind(this);
    this.onClickClear = this.onClickClear.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
  }

  onBlur = () => {
    this.props.onBlur();
  };

  onSearch(event, enterPress = false) {
    this.setState({
      value: event.target.value
    }, () => this.props.onSearch(this.state.value, enterPress));
  }

  onClickClear() {
    this.setState({
      value: '',
    }, () => this.props.onSearch(''));
  }

  onKeyPress(event) {
    if (shouldHandle(event)) {
      this.onSearch(event, true);
    }
  }

  render() {
    const {
      styleOverride, closeIcon, ...others
    } = this.props;

    const { value } = this.state;

    return (
      <div
        style={{
          ...searchBarStyle,
          ...this.style,
          ...styleOverride
        }}
      >
        <div style={searchBarStyle.searchIcon}>
          <SearchIcon width='12px' fill='#004c97' />
        </div>
        <div style={{ ...searchBarStyle.textFieldContainerStyle, ...this.style.container }}>
          <TextField
            name='search-bar'
            onChange={this.onSearch}
            defaultValue={value}
            placeholder='Find Conversations'
            onKeyPress={this.onKeyPress}
            onBlur={this.onBlur}
            styleOverride={{ ...searchBarStyle.textFieldStyle, ...styleOverride.textField }}
            {...others}
          />
        </div>
        {closeIcon && (
          <div style={{ ...searchBarStyle.clearIcon, ...styleOverride.clearIcon }}>
            <IconButton
              onClick={this.onClickClear}
              icon={Xmark}
              data-qa={'clear-topic'}
              padding={'12px 10px'}
              title='Clear'
            />
          </div>
        )}
      </div>
    );
  }
}

ConversationSearchBar.propTypes = {
  onBlur: PropTypes.func,
  onSearch: PropTypes.func,
  value: PropTypes.string,
  closeIcon: PropTypes.bool,
  styleOverride: PropTypes.func
};

ConversationSearchBar.defaultProps = {
  onBlur: () => {},
  onSearch: () => {},
  value: '',
  closeIcon: false,
  styleOverride: {},
};

export default ConversationSearchBar;
