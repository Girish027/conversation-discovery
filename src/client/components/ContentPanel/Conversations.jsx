import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Checkbox } from '@tfs/ui-components';
import EnhancedTable, { getColumnVisibility, getColumnWidth } from 'components/Table/EnhancedTable';
import { loadStorage } from '../../utils/storageManager';

export class Conversations extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedConversations: props.data,
    };
    this.tableIndentifier = 'CFD_Conversations_Table';
    this.tableInfo = {
      selected: {
        header: 'Checkbox',
        id: 'checkbox',
      },
      conversations: {
        header: 'Utterances',
        id: 'sentenceSet',
      },
    };
  }

  componentDidMount() {
    const { selectedConversations } = this.state;
    this.props.selectedConversations(selectedConversations);
  }

  get columns() {
    const { selectedConversations } = this.state;
    const tableState = loadStorage(this.tableIndentifier, {});

    const columns = [
      {
        Header: (row) => (
          <Checkbox
            styleOverride={{ paddingLeft: '10px' }}
            onChange={(event) => {
              this.toggleSelectAll(event.target.checked, row);
            }}
            checked={(selectedConversations.length > 0 && selectedConversations.length === this.props.data.length)}
          />
        ),
        id: 'checkbox',
        accessor: '',
        Cell: ({ original }) => (
          <Checkbox
            styleOverride={{ paddingLeft: '10px', marginTop: '5px' }}
            onChange={(event) => this.selectConversation(original, event.target.checked)}
            checked={selectedConversations.find((conversation) => conversation.transcriptId === original.transcriptId)}
          />
        ),
        sortable: false,
        maxWidth: 80,
        show: getColumnVisibility(tableState, this.tableInfo.selected.id, true),
        resizable: false,
      },
      {
        Header: this.tableInfo.conversations.header,
        accessor: this.tableInfo.conversations.id,
        sortable: true,
        minWidth: getColumnWidth(tableState, this.tableInfo.conversations.id, 80),
        show: getColumnVisibility(tableState, this.tableInfo.conversations.id, true),
      },
    ];

    return columns;
  }

  updateSelectedConversations = (selectedConversations) => {
    this.setState({
      selectedConversations,
    },
    this.props.selectedConversations(selectedConversations));
  };

  toggleSelectAll = (checked) => {
    let selectedConversations = [];
    if (checked) {
      selectedConversations = this.props.data;
    }
    this.updateSelectedConversations(selectedConversations);
  };

  selectConversation = (selectedConversation, checked) => {
    let { selectedConversations } = this.state;

    if (checked === true) {
      selectedConversations = [...selectedConversations, selectedConversation];
    } else {
      selectedConversations = selectedConversations.filter((conversation) => conversation.transcriptId !== selectedConversation.transcriptId);
    }
    this.updateSelectedConversations(selectedConversations);
  };

  render() {
    const { data } = this.props;

    return (
      <EnhancedTable
        tableIndentifier={this.tableIndentifier}
        columns={this.columns}
        data={data}
        tableStyle={{ maxHeight: '425px', maxWidth: '835x' }}
      />
    );
  }
}

Conversations.propTypes = {
  data: PropTypes.array,
  selectedConversations: PropTypes.func
};

Conversations.defaultProps = {
  data: [],
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});

export default connect(mapDispatchToProps)(Conversations);
