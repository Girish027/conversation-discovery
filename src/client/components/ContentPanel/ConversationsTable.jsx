import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Checkbox, Tooltip } from '@tfs/ui-components';
import EnhancedTable, { getColumnVisibility, getColumnWidth } from '../Table/EnhancedTable';
import { loadStorage } from '../../utils/storageManager';
import PreviewManager from '../../utils/PreviewManager';
import Preview from '../Preview/index';
import Constants from 'components/../constants';
import {
  setDefaultConversations
} from 'state/conversationsState';
import { conversationButtonStyle, conversationTableCheckBoxEnableStyle, conversationTableCheckBoxDisableStyle } from '../../styles/index';

export class ConversationsTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedConversations: this.props.selectedConvList,
    };

    this.tableIndentifier = 'CFD_Conversations_Table';
    this.tableInfo = {
      selected: {
        header: 'Checkbox',
        id: 'checkbox',
      },
      conversations: {
        header: 'Utterances',
        id: 'sentenceSet',
      },
      similarity: {
        header: 'Similarity',
        id: 'originalSimilarity',
      },
      intent: {
        header: 'Mapped To',
        id: 'assignedIntent',
      },
      faq: {
        header: 'Mapped To',
        id: 'assignedFaq',
      }
    };

    this.getProps = (state, rowInfo) => ({
      style: {
        backgroundColor: this.changeCellBackgroundColor(rowInfo),
      },
    });
  }

  componentDidMount() {
    this.setState({
      selectedConversations: this.props.selectedConvList
    });
  }

  componentWillReceiveProps(nextProps) {
    // Any time props.selectedAction changes, update state selectedConversations to empty list.
    if (nextProps.selectedAction !== this.props.selectedAction) {
      this.setState({
        selectedConversations: []
      });
    }
  }

  componentWillUnmount() {
    PreviewManager.closeWidget();
  }

  getCheckBoxHeaderChecked = () => {
    const { selectedConversations } = this.state;
    const {
      isAnswers, isConversations, selectedAction
    } = this.props;

    return (selectedConversations.length > 0 && ((isAnswers && isConversations) ? ((selectedAction === Constants.utteranceMode.conversations)
      ? ((this.props.data.filter((conversation) => conversation.assignedIntent === '').length) === selectedConversations.length) : ((this.props.data.filter((conversation) => conversation.assignedFaq === '').length) === selectedConversations.length))
      : ((isConversations && !isAnswers) ? ((this.props.data.filter((conversation) => conversation.assignedIntent === '').length) === selectedConversations.length) : ((this.props.data.filter((conversation) => conversation.assignedFaq === '').length) === selectedConversations.length))));
  }

  getCheckBoxCellStyle = (original) => {
    const {
      isAnswers, isConversations, selectedAction
    } = this.props;

    return (isAnswers && isConversations) ? ((selectedAction === Constants.utteranceMode.conversations)
      ? (original.assignedIntent === '' ? conversationTableCheckBoxEnableStyle : conversationTableCheckBoxDisableStyle) : (original.assignedFaq === '' ? conversationTableCheckBoxEnableStyle : conversationTableCheckBoxDisableStyle))
      : ((isConversations && !isAnswers) ? (original.assignedIntent === '' ? conversationTableCheckBoxEnableStyle : conversationTableCheckBoxDisableStyle) : (original.assignedFaq === '' ? conversationTableCheckBoxEnableStyle : conversationTableCheckBoxDisableStyle));
  }

  get columns() {
    const { selectedConversations } = this.state;
    const {
      isAnswers, isConversations, selectedAction, roleConfig, dispatch
    } = this.props;
    if (selectedConversations.length <= 0) {
      dispatch(setDefaultConversations());
    }
    const tableState = loadStorage(this.tableIndentifier, {});
    const mappingColumnAccessor = (isAnswers && isConversations) ? ((selectedAction === Constants.utteranceMode.conversations)
      ? this.tableInfo.intent.id : this.tableInfo.faq.id)
      : ((!isAnswers && isConversations) ? this.tableInfo.intent.id : this.tableInfo.faq.id);

    const columns = [
      {
        Header: () => (
          <Checkbox
            styleOverride={{ paddingLeft: '10px' }}
            onChange={(event) => {
              this.toggleSelectAll(event.target.checked);
            }}
            checked={this.getCheckBoxHeaderChecked()}
          />
        ),
        id: 'checkbox',
        accessor: '',
        Cell: ({ original }) => (
          <Checkbox
            styleOverride={this.getCheckBoxCellStyle(original)}
            onChange={(event) => this.selectConversation(original, event.target.checked)}
            checked={selectedConversations.find((conversation) => conversation.transcriptId === original.transcriptId)}
          />
        ),
        sortable: false,
        maxWidth: 80,
        show: getColumnVisibility(tableState, this.tableInfo.selected.id, (roleConfig.addToConversations && roleConfig.addToAnswers)),
        resizable: false,
        getProps: this.getProps,
      },
      {
        Header: this.tableInfo.conversations.header,
        accessor: this.tableInfo.conversations.id,
        sortable: true,
        minWidth: getColumnWidth(tableState, this.tableInfo.conversations.id, 380),
        show: getColumnVisibility(tableState, this.tableInfo.conversations.id, true),
        getProps: this.getProps,
        Cell: ({ original }) => (

          <button type='button' style={conversationButtonStyle} onClick={() => this.props.getTranscript(original.transcriptId)}>
            <Tooltip content='Click to view full transcript' type='info' direction='bottom'>
              <div>
                {original.sentenceSet}
              </div>
            </Tooltip>
          </button>

        )
      },
      {
        Header: this.tableInfo.similarity.header,
        accessor: this.tableInfo.similarity.id,
        sortable: true,
        minWidth: getColumnWidth(tableState, this.tableInfo.similarity.id, 100),
        show: getColumnVisibility(tableState, this.tableInfo.similarity.id, true),
        Cell: ({ original }) => (
          <span>
            {' '}
            { original.originalSimilarity.toFixed(4) }
            {' '}
          </span>
        ),
        getProps: this.getProps,
      },
      {
        Header: this.tableInfo.intent.header,
        accessor: mappingColumnAccessor,
        sortable: true,
        minWidth: getColumnWidth(tableState, this.tableInfo.intent.id, 180),
        show: getColumnVisibility(tableState, this.tableInfo.intent.id, true),
        Cell: ({ original }) => (
          <span>
            { (isConversations && original.assignedIntent !== '') ? (
              <div>
                {' '}
Intent :
                {original.assignedIntent}
              </div>
            ) : (<div></div>)}
            {
              (isAnswers && original.assignedFaq !== '') ? (
                <div>
                  {' '}
  FAQ :
                  {original.assignedFaq}
                </div>
              ) : (<div></div>)
            }
          </span>
        ),
        getProps: this.getProps,
      },
    ];

    return columns;
  }

  changeCellBackgroundColor = (rowInfo) => {
    const { isAnswers, isConversations, roleConfig } = this.props;
    const roleAccess = (roleConfig.addToConversations && roleConfig.addToAnswers);
    return ((isAnswers && isConversations)
      ? ((rowInfo && rowInfo.row.assignedIntent !== '' && rowInfo.row.assignedFaq !== '' && roleAccess)
        ? '#A8A8A8' : null)
      : ((!isAnswers && isConversations)
        ? ((rowInfo && rowInfo.row.assignedIntent !== '' && roleAccess)
          ? '#A8A8A8' : null)
        : ((rowInfo && rowInfo.row.assignedFaq !== '' && roleAccess)
          ? '#A8A8A8' : null)));
  }

  updateSelectedConversations = (selectedConversations) => {
    this.setState({
      selectedConversations,
    },
    this.props.selectedConversations(selectedConversations));
  };

  toggleSelectAll = (checked) => {
    const { isAnswers, isConversations, selectedAction } = this.props;
    let selectedConversations = [];
    if (checked) {
      selectedConversations = this.props.data;
      if (isAnswers && isConversations) {
        if (selectedAction === Constants.utteranceMode.conversations) {
          selectedConversations = selectedConversations.filter((conversation) => conversation.assignedIntent === '');
        } else {
          selectedConversations = selectedConversations.filter((conversation) => conversation.assignedFaq === '');
        }
      } else if (isConversations) {
        selectedConversations = selectedConversations.filter((conversation) => conversation.assignedIntent === '');
      } else {
        selectedConversations = selectedConversations.filter((conversation) => conversation.assignedFaq === '');
      }
    }
    this.updateSelectedConversations(selectedConversations);
  };

  selectConversation = (selectedConversation, checked) => {
    let { selectedConversations } = this.state;

    if (checked === true) {
      selectedConversations = [...selectedConversations, selectedConversation];
    } else {
      selectedConversations = selectedConversations.filter((conversation) => conversation.transcriptId !== selectedConversation.transcriptId);
    }
    this.updateSelectedConversations(selectedConversations);
  };

  render() {
    const { data } = this.props;

    return (
      <div>
        <EnhancedTable
          tableIndentifier={this.tableIndentifier}
          columns={this.columns}
          data={data}
          tableStyle={{ maxHeight: 'calc(100vh - 215px)' }}
          defaultSorted={[{ id: 'originalSimilarity', desc: true }]}
          defaultPageSize={30}
        />
        <Preview />
      </div>
    );
  }
}

ConversationsTable.propTypes = {
  getTranscript: PropTypes.func,
  dispatch: PropTypes.func,
  data: PropTypes.array,
  selectedConversations: PropTypes.func,
  selectedConvList: PropTypes.array,
  selectedAction: PropTypes.string,
  isAnswers: PropTypes.bool,
  isConversations: PropTypes.bool,
  roleConfig: PropTypes.object,
};

ConversationsTable.defaultProps = {
  dispatch: () => {},
  data: [],
  selectedConvList: [],
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});
export default connect(mapDispatchToProps)(ConversationsTable);
