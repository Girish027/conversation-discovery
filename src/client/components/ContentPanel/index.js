import { connect } from 'react-redux';
import ContentPanel from './ContentPanel';
import { getPathnameSelector } from 'state/routerState';

const mapStateToProps = (state) => ({
  routeName: getPathnameSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});


export default connect(mapStateToProps, mapDispatchToProps)(ContentPanel);
