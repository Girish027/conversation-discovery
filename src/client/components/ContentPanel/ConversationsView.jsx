import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  List, Map
} from 'immutable';
import _ from 'lodash';
import Placeholder from 'components/Placeholder';
import {
  Pencil,
  Button,
  ContextualActionItem,
  Checkmark,
} from '@tfs/ui-components';
import { runActions, runSelectors } from 'state/runsState';
import {
  clusterSelectors, clusterActions
} from 'state/clusterState';
import {
  conversationSelectors, conversationActions, setDefaultConversations
} from 'state/conversationsState';
import Constants from 'components/../constants';
import IconButton from 'components/IconButton';
import LoadingIndicator from 'components/LoadingIndicator';
import KeywordsSearchBar from '../KeywordsSearchBar';
import { clusterAction } from '../../styles/index';
import { ConversationsTable } from './ConversationsTable';
import {
  setModalIsOpen, getIsConversationsSelector, getIsAnswersSelector, getRoleConfigSelector
} from '../../state/appState';
import { windowStyleConversationView } from 'styles';
import { prepareClusterName } from '../../utils/stringOperations';

export class ConversationsView extends Component {
  constructor(props) {
    super(props);
    this.state = {
      cluster: {},
      conversations: [],
      selectedConvList: [],
      searchedConversation: '',
      filteredItems: {},
      selectedAction: Constants.utteranceMode.conversations,
      modeAnswers: false
    };
  }

  static getDerivedStateFromProps(props) {
    const { cluster, conversations } = props;
    return {
      cluster: cluster.toJS(),
      conversations: conversations.toJS(),
    };
  }

  componentDidMount() {
    const { dispatch, isAnswers, isConversations } = this.props;
    if (isConversations) {
      dispatch(conversationActions.getIntents());
    }
    if (isAnswers) {
      dispatch(conversationActions.getInterfaces());
    }
    dispatch(conversationActions.getAllConversations());
  }

  // gets transcript for a particular conversation
  getTranscript = (transcriptId) => {
    const { dispatch } = this.props;
    dispatch(conversationActions.getConversationTranscript(transcriptId));
  }

  goToRunOverview = () => {
    const { selectedRunId, dispatch } = this.props;
    dispatch(runActions.selectRun(selectedRunId));
  };

  // sample code for testing Edit on cluster
  onEditCluster = (cluster) => {
    const { dispatch } = this.props;

    const data = {
      clusterName: cluster.clusterName,
      clusterDescription: cluster.clusterDescription
    };

    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.editCluster,
      clusterId: cluster.clusterId,
      formData: data,
    }));
  };

  resetConversations = (utterances) => {
    const utt = utterances || [];
    const data = this.state.selectedConvList;
    if ((data.length === 0) && (utt.length !== 0)) {
      utt.forEach((element) => {
        const obj = {};
        obj.sentenceSet = element.utterance;
        obj.transcriptId = element.utteranceId;
        data.push(obj);
      });
    }
    return data;
  }

  onClickAddToBot = () => {
    const { dispatch, selectedConversation } = this.props;
    const utterances = (selectedConversation) ? selectedConversation.utterances : [];
    const data = this.resetConversations(utterances);
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.addToBot,
      data,
      selectedConversations: (selectedConversations) => this.selectedConversations(selectedConversations)
    }));
  };

  onClickAddToFAQ = () => {
    const { dispatch, selectedConversation } = this.props;
    const utterances = (selectedConversation) ? selectedConversation.utterances : [];
    const data = this.resetConversations(utterances);
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.addToFaq,
      data,
      selectedConversations: (selectedConversations) => this.selectedConversations(selectedConversations)
    }));
  };

  onSearch = (searchedConversation, enterPressed) => {
    const { dispatch } = this.props;
    this.setState({
      searchedConversation
    });

    let searchItems;
    if (searchedConversation !== '') {
      searchItems = { ...this.state.filteredItems, search: searchedConversation };
    } else {
      searchItems = { ...this.state.filteredItems };
    }
    if (enterPressed) {
      dispatch(conversationActions.getAllConversations(searchItems));
    }
  }

  onFilter = (filteredItems) => {
    const { dispatch } = this.props;
    this.setState({
      filteredItems
    });
    let searchItems;
    if (filteredItems.similarity != null) {
      if (this.state.searchedConversation !== '') {
        searchItems = { search: this.state.searchedConversation, similarity: filteredItems.similarity };
      } else {
        searchItems = { similarity: filteredItems.similarity };
      }
      dispatch(conversationActions.getAllConversations(searchItems));
    } else {
      if (this.state.searchedConversation !== '') {
        searchItems = { search: this.state.searchedConversation };
      }
      if (_.isEmpty(searchItems)) {
        dispatch(conversationActions.getAllConversations());
      } else {
        dispatch(conversationActions.getAllConversations(searchItems));
      }
    }
  }

  onClickMarkAsReviewed = () => {
    const { cluster } = this.state;
    const { dispatch } = this.props;
    const data = {
      type: 'finalize'
    };
    dispatch(clusterActions.editCluster(cluster.clusterId, data));
    window.location.reload();
  };

  handleModeChange = () => {
    const { dispatch } = this.props;
    const { modeAnswers } = this.state;
    const selectedMode = (!modeAnswers) ? Constants.utteranceMode.answers : Constants.utteranceMode.conversations;
    dispatch(setDefaultConversations());
    this.setState((prevState) => ({
      modeAnswers: !prevState.modeAnswers,
      selectedConvList: [],
      selectedAction: selectedMode
    }));
  };

  renderClusterHeaderBar = () => {
    const { cluster, selectedConvList, modeAnswers } = this.state;
    const { clusterName } = cluster;
    const {
      isAnswers, isConversations, selectedConversation, roleConfig
    } = this.props;
    const utteranceList = (selectedConversation === undefined) ? [] : selectedConversation.utterances;
    const utterLength = (utteranceList) ? utteranceList.length : 0;
    return (
      <div
        style={clusterAction.renameIconLabel}
      >
        <div
          className='cluster-header'
          style={clusterAction.renameIconLabel.children}
        >
          <div style={clusterAction.renameIconLabel.clusterName}>
            <span
              role='button'
              tabIndex={0}
            >
              {prepareClusterName(clusterName)}
            </span>
            { roleConfig.editPotentialIntentsNames
              ? (
                <IconButton
                  icon={Pencil}
                  onClick={() => this.onEditCluster(cluster)}
                  defaultColor='none'
                  strokeColor='#004c97'
                  height={13}
                  width={13}
                  padding={'5px'}
                />
              ) : null }
          </div>
        </div>

        <div
          className='add-to-container'
          style={{
          }}
        >
          { roleConfig.markAsReviewed
            ? (
              <Button
                styleOverride={clusterAction.buttons}
                type='flat'
                name='run-overview'
              >
                { !this.state.cluster.finalized ? (
                  <ContextualActionItem
                    icon={Checkmark}
                    onClickAction={this.onClickMarkAsReviewed}
                    styleOverride={{
                      marginTop: '0px',
                      paddingLeft: '10px',
                    }}
                  >
                MARK AS REVIEWED
                  </ContextualActionItem>
                ) : (
                  <ContextualActionItem
                    icon={Checkmark}
                    disabled
                    styleOverride={{
                      color: '#c6c6c6',
                      marginTop: '0px',
                      paddingLeft: '10px',
                    }}
                  >
                  REVIEWED
                  </ContextualActionItem>
                )
                }
              </Button>
            ) : null }
          {(isAnswers && isConversations && roleConfig.addToConversations && roleConfig.addToAnswers) ? (
            <div style={clusterAction.modeSwitch}>
              <Button
                name='conversation-mode'
                type='flat'
                styleOverride={
                  (modeAnswers)
                    ? { ...clusterAction.modeSwitch.conversations, ...clusterAction.modeSwitch.off }
                    : { ...clusterAction.modeSwitch.conversations, ...clusterAction.modeSwitch.on }
                }
                onClick={(modeAnswers) ? this.handleModeChange : () => { }}
              >
                {Constants.utteranceMode.conversations}
              </Button>
              <Button
                name='answers-mode'
                type='flat'
                styleOverride={
                  (modeAnswers)
                    ? { ...clusterAction.modeSwitch.answers, ...clusterAction.modeSwitch.on }
                    : { ...clusterAction.modeSwitch.answers, ...clusterAction.modeSwitch.off }
                }
                onClick={(modeAnswers) ? () => { } : this.handleModeChange}
              >
                {Constants.utteranceMode.answers}
              </Button>
            </div>
          ) : null}
          { (roleConfig.addToAnswers && roleConfig.addToConversations)
            ? (
              <Button
                styleOverride={clusterAction.buttons.lastChild}
                type='primary'
                disabled={selectedConvList.length === 0 && utterLength === 0}
                name='add-bot'
                onClick={(isAnswers && isConversations) ? ((this.state.modeAnswers)
                  ? this.onClickAddToFAQ : this.onClickAddToBot)
                  : ((isConversations && !isAnswers) ? this.onClickAddToBot : this.onClickAddToFAQ)}
              >
                {(isAnswers && isConversations) ? ((this.state.modeAnswers)
                  ? Constants.buttons.addToBotIntent : Constants.buttons.addToBotIntent)
                  : ((isConversations && !isAnswers) ? Constants.buttons.addToBotIntent : Constants.buttons.addToBotIntent)}
              </Button>
            ) : null }
        </div>
      </div>
    );
  }

  renderKeywardSearchBar = () => {
    const { cluster } = this.state;
    const { wordCloudTerms } = cluster;
    let quickSearchList = [];
    if (wordCloudTerms) {
      const quickSearch = wordCloudTerms.split(/[_,]+/);
      quickSearchList = quickSearch.filter((item, pos, self) => self.indexOf(item) === pos);
    }

    return (
      <div
        style={clusterAction.quickSearchHeader}
      >
        <KeywordsSearchBar
          words={quickSearchList}
          onSearch={(conversation, keyCode) => this.onSearch(conversation, keyCode)}
          onBlur={() => {}}
        />
      </div>
    );
  }

  renderClusterActions = () => (
    <div>
      {this.renderClusterHeaderBar()}
      {this.renderKeywardSearchBar()}
    </div>
  )

  renderPlaceholder = () => (
    <Placeholder>
      <div className='message-default'>
        <LoadingIndicator isHerbie message='Loading Conversations...' />
      </div>
    </Placeholder>
  );

  selectedConversations = (selectedConversations) => {
    this.setState({
      selectedConvList: selectedConversations,
    });
  }

  render() {
    const { conversations, selectedConvList } = this.state;
    const { conversationCount, dispatch } = this.props;

    if (conversationCount === -1) {
      return this.renderPlaceholder();
    }
    return (
      <div>
        <div style={windowStyleConversationView}>
          {this.renderClusterActions()}
        </div>
        <ConversationsTable
          data={conversations}
          selectedAction={this.state.selectedAction}
          isAnswers={this.props.isAnswers}
          isConversations={this.props.isConversations}
          roleConfig={this.props.roleConfig}
          getTranscript={this.getTranscript}
          selectedConvList={selectedConvList}
          selectedConversations={(selectedConversations) => this.selectedConversations(selectedConversations)}
          dispatch={dispatch}
        />
      </div>
    );
  }
}

ConversationsView.propTypes = {
  selectedRunId: PropTypes.string,
  dispatch: PropTypes.func,
  // Note: https://github.com/yannickcr/eslint-plugin-react/issues/1751
  // eslint-disable-next-line react/no-unused-prop-types
  cluster: PropTypes.object,
  // eslint-disable-next-line react/no-unused-prop-types
  conversations: PropTypes.object,
  conversationCount: PropTypes.number,
  isAnswers: PropTypes.bool,
  isConversations: PropTypes.bool,
  selectedConversation: PropTypes.object,
  roleConfig: PropTypes.object,
};

ConversationsView.defaultProps = {
  dispatch: () => { },
  selectedRunId: '',
  cluster: Map(),
  conversations: new List([]),
  conversationCount: -1,
  isConversations: false,
  selectedConversation: new List([]),
  roleConfig: {},
};


const mapStateToProps = (state) => ({
  selectedRunId: runSelectors.getSelectedRunId(state),
  cluster: clusterSelectors.getSelectedCluster(state),
  conversations: conversationSelectors.getConversationsList(state),
  conversationCount: conversationSelectors.getConversationCount(state),
  isAnswers: getIsAnswersSelector(state),
  isConversations: getIsConversationsSelector(state),
  selectedConversation: conversationSelectors.getSelectedConversations(state),
  roleConfig: getRoleConfigSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});

export default connect(mapStateToProps, mapDispatchToProps)(ConversationsView);
