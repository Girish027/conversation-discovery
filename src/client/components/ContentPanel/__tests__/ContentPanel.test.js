import React from 'react';
import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { fromJS } from 'immutable';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import ConnectedContentPanel from 'components/ContentPanel';
import ContentPanel from 'components/ContentPanel/ContentPanel';
import { routeNames } from 'utils/RouteHelper';

describe('<ContentPanel />', () => {
  let wrapper;
  let store;

  const caa = {
    clientId: '247inc',
    appId: 'referencebot',
    accountId: 'referencebot',
  };

  const props = {
    caa,
    routeName: routeNames.DISCOVER_INTENTS,
  };


  beforeAll(() => {
    // Mock store
    const mockStore = configureStore();
    store = mockStore(fromJS({
      router: {
        location: {
          pathname: routeNames.DISCOVER_INTENTS
        }
      },
      app: {
        ...caa
      },
      projects: {
        activeProjectId: 'pro-123',
        projects: [{projectId: 'pro-123', projectType: 'SYSTEM'}]
      },
      runs: {
        runs: []
      }

    }));
  });


  afterAll(() => {
    jest.clearAllMocks();
  });

  const getWrapper = () => shallow(
    <ContentPanel {...props} />
  );

  const getConnectedWrapper = () => mount(
    <Provider store={store}>
      <ConnectedContentPanel />
    </Provider>);

  describe('Creating an instance', () => {
    test('should exist - connected components', () => {
      wrapper = getConnectedWrapper();
      expect(wrapper.exists()).toBe(true);
    });

    test('should exist - basic', () => {
      wrapper = getWrapper();
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with Runs View', () => {
      wrapper = shallow(<ContentPanel {...props} />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
    test('renders correctly with Topic Distribution View', () => {
      wrapper = shallow(<ContentPanel routeName={routeNames.DISTRIBUTION_GRAPH} />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
    test('renders correctly with Conversations View', () => {
      wrapper = shallow(<ContentPanel routeName={routeNames.TOPIC_REVIEW} />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
    test('renders correctly with Runs View', () => {
      wrapper = shallow(<ContentPanel routeName={routeNames.BASE_ROUTE} />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
    test('renders correctly with Runs View', () => {
      wrapper = shallow(<ContentPanel routeName={routeNames.DISCOVER_INTENTS} />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
