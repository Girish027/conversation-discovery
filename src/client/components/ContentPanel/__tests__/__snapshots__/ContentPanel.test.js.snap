// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentPanel /> Snapshots renders correctly with Conversations View 1`] = `<Connect(ConversationsView) />`;

exports[`<ContentPanel /> Snapshots renders correctly with Runs View 1`] = `<Connect(RunsView) />`;

exports[`<ContentPanel /> Snapshots renders correctly with Runs View 2`] = `<Connect(RunsView) />`;

exports[`<ContentPanel /> Snapshots renders correctly with Runs View 3`] = `<Connect(RunsView) />`;

exports[`<ContentPanel /> Snapshots renders correctly with Topic Distribution View 1`] = `<Connect(TopicDistributionView) />`;
